src/net/tcp/stream.rs

```diff
@@ -149,6 +167,24 @@ impl ReadHalf {
             ))),
         }
     }
+
+    /// Put bytes in `buf` based on the minimum of `avail` and its remaining
+    /// capacity.
+    ///
+    /// Returns an optional `Bytes` containing any remainder of `avail` that was
+    /// not consumed.
+    fn put_slice(avail: Bytes, buf: &mut ReadBuf) -> Option<Bytes> {
+        let amt = std::cmp::min(avail.len(), buf.remaining());
+        let (a, b) = avail.split_at(amt);
+
+        buf.put_slice(a);
+
+        if b.is_empty() {
+            None
+        } else {
+            Some(Bytes::copy_from_slice(b))
+        }
```

Comment:
Better to use `advance` so you don't have to copy the bytes out:

```suggestion
        if amt == 0 {
            return None;
        }
        
        buf.put_slice(&avail[..amt]);
        avail.advance(amt);

        if avail.is_empty() {
            None
        } else {
            Some(avail)
        }
```